src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java

```diff
@@ -0,0 +1,253 @@
+  public static final String NAME = "BigQueryExecute";
+  private static final String RECORDS_PROCESSED = "records.processed";
+
+  private Config config;
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    config.validate(context.getFailureCollector());
+    QueryJobConfiguration.Builder builder = QueryJobConfiguration.newBuilder(config.getSql());
+    // Run at batch priority, which won't count toward concurrent rate limit.
+    if (config.getMode().equals(QueryJobConfiguration.Priority.BATCH)) {
+      builder.setPriority(QueryJobConfiguration.Priority.BATCH);
+    } else {
+      builder.setPriority(QueryJobConfiguration.Priority.INTERACTIVE);
+    }
+
+    // Save the results of the query to a permanent table.
+    if (config.getDataset() != null && config.getTable() != null) {
+      builder.setDestinationTable(TableId.of(config.getDataset(), config.getTable()));
+    }
+
+    // Enable or Disable the query cache to force live query evaluation.
+    if (config.shouldUseCache()) {
+      builder.setUseQueryCache(true);
+    }
+
+    // Enable legacy SQL
+    builder.setUseLegacySql(config.isLegacySQL());
+
+    QueryJobConfiguration queryConfig = builder.build();
+
+    // Location must match that of the dataset(s) referenced in the query.
+    JobId jobId = JobId.newBuilder().setRandomJob().setLocation(config.getLocation()).build();
+
+    // API request - starts the query.
+    Credentials credentials = config.getServiceAccountFilePath() == null ?
+                                null : GCPUtils.loadServiceAccountCredentials(config.getServiceAccountFilePath());
+    BigQuery bigQuery = GCPUtils.getBigQuery(config.getProject(), credentials);
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+
+    LOG.info("Executing SQL as job {}.", jobId.getJob());
+    LOG.debug("The BigQuery SQL is {}", config.getSql());
+
+    // Wait for the query to complete
+    queryJob.waitFor();
+
+    // Check for errors
+    if (queryJob.getStatus().getError() != null) {
+      // You can also look at queryJob.getStatus().getExecutionErrors() for all
+      // errors, not just the latest one.
+      throw new RuntimeException(queryJob.getStatus().getExecutionErrors().toString());
+    }
+
+    TableResult queryResults = queryJob.getQueryResults();
+    long rows = queryResults.getTotalRows();
+    if (config.shouldSetAsArguments()) {
+      if (rows == 0 || queryResults.getSchema() == null) {
+        LOG.warn("The query result does not contain any row or schema, will not save the results in the arguments");
+      } else {
+        Schema schema = queryResults.getSchema();
+        FieldValueList firstRow = queryResults.iterateAll().iterator().next();
+        for (int i = 0; i < schema.getFields().size(); i++) {
+          Field field = schema.getFields().get(i);
+          String name = field.getName();
+          if (field.getType().equals(LegacySQLTypeName.RECORD)) {
```

Comment:
what about arrays?

Reply:
They don't have array as the type, from https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/LegacySQLTypeName.java#L89, the only nested type is RECORD

Reply:
bigquery array is represented as repeated mode of some type.

Reply:
Updated the check for mode